File: src/ios/FirebaseCrashPlugin.m
#import "FirebaseCrashPlugin.h"

//@import Firebase;
@import FirebaseCrashlytics;
@import FirebaseCore;

@implementation FirebaseCrashPlugin

- (void)pluginInitialize {
    NSLog(@"Starting Firebase Crashlytics plugin");

    if(![FIRApp defaultApp]) {
        [FIRApp configure];
    }
}

- (void)log:(CDVInvokedUrlCommand *)command {
    NSString* errorMessage = [command.arguments objectAtIndex:0];

    [[FIRCrashlytics crashlytics] log:errorMessage];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setCustomKeys:(CDVInvokedUrlCommand *)command {
    NSString* keyname1 = [command.arguments objectAtIndex:0];
    NSString* keyvalue1 = [command.arguments objectAtIndex:1];
    NSString* keyname2 = [command.arguments objectAtIndex:2];
    NSString* keyvalue2 = [command.arguments objectAtIndex:3];
    NSString* keyname3 = [command.arguments objectAtIndex:4];
    NSString* keyvalue3 = [command.arguments objectAtIndex:5];
    NSString* keyname4 = [command.arguments objectAtIndex:6];
    NSString* keyvalue4 = [command.arguments objectAtIndex:7];

    @try{  
        NSMutableDictionary *keysAndValues = [NSMutableDictionary dictionary];
        
        if (![keyname1 isEqualToString:@""] && ![keyvalue1 isEqualToString:@""]) {
            [keysAndValues setObject:keyvalue1 forKey:keyname1];
        }
        if (![keyname2 isEqualToString:@""] && ![keyvalue2 isEqualToString:@""]) {
            [keysAndValues setObject:keyvalue2 forKey:keyname2];
        }
        if (![keyname3 isEqualToString:@""] && ![keyvalue3 isEqualToString:@""]) {
            [keysAndValues setObject:keyvalue3 forKey:keyname3];
        }
        if (![keyname4 isEqualToString:@""] && ![keyvalue4 isEqualToString:@""]) {
            [keysAndValues setObject:keyvalue4 forKey:keyname4];
        }
    
        if ([keysAndValues count] > 0) {
            [[FIRCrashlytics crashlytics] setCustomKeysAndValues:keysAndValues];
        }
    
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }@catch (NSException* exception) {
        CDVPluginResult* pluginResultErr = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[exception reason]];  
        [self.commandDelegate sendPluginResult:pluginResultErr callbackId:command.callbackId];
    }
}

- (void)logError:(CDVInvokedUrlCommand*)command {
    NSString *description = NSLocalizedString([command argumentAtIndex:0 withDefault:@"No Message Provided"], nil);
    NSDictionary *userInfo = @{ NSLocalizedDescriptionKey: description };
    NSNumber *defaultCode = [NSNumber numberWithInt:-1];
    int code = [[command argumentAtIndex:1 withDefault:defaultCode] intValue];
    NSString *domain = [[NSBundle mainBundle] bundleIdentifier];
    NSError *error = [NSError errorWithDomain:domain code:code userInfo:userInfo];

    [[FIRCrashlytics crashlytics] recordError:error];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setUserId:(CDVInvokedUrlCommand *)command {
    NSString* userId = [command.arguments objectAtIndex:0];

    [[FIRCrashlytics crashlytics] setUserID:userId];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setEnabled:(CDVInvokedUrlCommand *)command {
    bool enabled = [[command.arguments objectAtIndex:0] boolValue];

    [[FIRCrashlytics crashlytics] setCrashlyticsCollectionEnabled:enabled];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}


- (void)forceCrash:(CDVInvokedUrlCommand *)command {
    
    @[][1];
    
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}



@end

File: src/android/FirebaseCrashPlugin.java
package by.chemerisuk.cordova.firebase;

import android.util.Log;

import by.chemerisuk.cordova.support.CordovaMethod;
import by.chemerisuk.cordova.support.ReflectiveCordovaPlugin;

import com.google.firebase.crashlytics.FirebaseCrashlytics;

import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONException;


public class FirebaseCrashPlugin extends ReflectiveCordovaPlugin {
    private final String TAG = "FirebaseCrashPlugin";

    private FirebaseCrashlytics firebaseCrashlytics;

    @Override
    protected void pluginInitialize() {
        Log.d(TAG, "Starting Firebase Crashlytics plugin");

        try{
            firebaseCrashlytics = FirebaseCrashlytics.getInstance();
        }
        catch (Exception e){
            Log.e("Firebase Crashlytics Plugin", e.getMessage());
        }
    }

    @CordovaMethod(ExecutionThread.WORKER)
    private void forceCrash(CallbackContext callbackContext) {
            new Thread(){
                @Override
                public void run() {
                    throw new RuntimeException("MyCrash");
                }
            }.start();
    }

    @CordovaMethod(ExecutionThread.WORKER)
    private void log(String message, CallbackContext callbackContext) {
        if(firebaseCrashlytics != null){
            firebaseCrashlytics.log(message);
            callbackContext.success();
        }
    }

    @CordovaMethod(ExecutionThread.UI)
    private void logError(String message, CallbackContext callbackContext) {
        try {
            if(firebaseCrashlytics != null){
                firebaseCrashlytics.recordException(new Exception(message));
                callbackContext.success();
            }
        } catch (Exception e) {
        }
    }

    @CordovaMethod(ExecutionThread.UI)
    private void setCustomKeys(String keyname1, String keyvalue1, String keyname2, String keyvalue2, String keyname3, String keyvalue3, String keyname4, String keyvalue4, CallbackContext callbackContext) {
        try {
            if(firebaseCrashlytics != null){        
                if(!keyname1.isEmpty() && !keyvalue1.isEmpty()){
                    firebaseCrashlytics.setCustomKey(keyname1, keyvalue1);
                }
                if(!keyname2.isEmpty() && !keyvalue2.isEmpty()){
                    firebaseCrashlytics.setCustomKey(keyname2, keyvalue2);
                }
                if(!keyname3.isEmpty() && !keyvalue3.isEmpty()){
                    firebaseCrashlytics.setCustomKey(keyname3, keyvalue3);
                }
                if(!keyname4.isEmpty() && !keyvalue4.isEmpty()){
                    firebaseCrashlytics.setCustomKey(keyname4, keyvalue4);
                }
                
                callbackContext.success();
            }
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
        }
    }

    @CordovaMethod(ExecutionThread.UI)
    private void setUserId(String userId, CallbackContext callbackContext) {
        if(firebaseCrashlytics != null){
            firebaseCrashlytics.setUserId(userId);
            callbackContext.success();
        }
    }

    @CordovaMethod
    private void setEnabled(boolean enabled, CallbackContext callbackContext) {
        if(firebaseCrashlytics != null){
            firebaseCrashlytics.setCrashlyticsCollectionEnabled(enabled);
            callbackContext.success();
        }
    }

}
